### ISSUE TITLE
New CVal conversion operator for native integer conversions

### ISSUE THREAD

**Issue by AnthonyDGreen** [dotnet/vblang#85]
This idea addresses scenarios #82 and #83.

**Comment by KlausLoeffelmann** [comment]
How would this operator work? Like Ctype?

**Comment by AnthonyDGreen** [comment]
@KlausLoeffelmann I updated the description with a little more detail. That really should go into a speclet so that we can update it with PRs.

**Comment by Bill-McC** [comment]
Boolean to Int.... hmmmm.  I'm sure I've heard that discussion (in length) before. Unless there is a serious performance differential in favour of the single bit, I'd say stick with all bits on/off.

**Comment by Bill-McC** [comment]
Re overflow checking: I thought the idea of cval was to avoid such things. Without that hard to see the difference between conversion operators.

If we get unchecked blocks,  I'd expect CVal to work the same regardless. I'd even like ChrW to avoid bound checking if inside an unchecked block, but then again that'd mean you'd have to exit the unchecked block to get traditional chrw.

So I think I'd prefer conversion operators to remain consistent,  with CVal focusing on high performance non bound check conversions, always.

**Comment by Nukepayload2** [comment]
Conversions between Integer and Boolean should behave like System.BitConverter .
```vb
Dim booleanValue = False
' Dim int32Value = CVal(booleanValue, Integer)
Dim int32Value As Integer = BitConverter.GetBytes(booleanValue)(0)
```
```vb
Dim int32Value = 1
' Dim booleanValue = CVal(int32Value, Boolean)
Dim booleanValue = BitConverter.ToBoolean(BitConverter.GetBytes(int32Value), 0)
```

**Comment by Bill-McC** [comment]
@Nukepayload2  What of int32Value =  255, or -1 etc.  They'd be true wouldn't they, as would 3, 5, 7, 9,  etc, but not 2, 4, 6, 8 ???

**Comment by Bill-McC** [comment]
@Nukepayload2  the BitConverter.ToBoolean( bytes(), startindex) method compares each byte to zero and it if it non zero returns True. So rather than convert the integer to a byte array, it'd be simpler to write:
`bool = If(value = 0, False, True)`
In either case any non zero value is True. Logically:  True =  Not False.

So using bitconverter or the above if expression, 1, 2, 3, 4, 5, etc are all True.

**Comment by Bill-McC** [comment]
Sorry, I should have properly tested BitConverter.ToBoolean(bytes(), startindex)
It only tests the byte at startindex. So values 1 to 255, return true, but values 256 * n return False. I.E, 
254 is True, 
255 is True, 
256 is False,
257 is True,
...
512 is False,
513 is True,

or basically it is testing just `value And &HFF <> 0`

**Comment by Nukepayload2** [comment]
The BitConverter.ToBoolean(Byte(), Integer) is testing value(0) <> 0.
```VB.NET
'==================================ToBoolean===================================
'        **Action:  Convert an array of bytes to a boolean value.  We treat this array 
'        **         as if the first 4 bytes were an Int4 an operate on this value.
'        **Returns: True if the Int4 value of the first 4 bytes is non-zero.
'        **Arguments: value -- The byte array
'        **           startIndex -- The position within the array.
'        **Exceptions: See ToInt4.
'        ==============================================================================

' Converts an array of bytes into a boolean.  
Public Shared Function ToBoolean(value As Byte(), startIndex As Integer) As Boolean
	If value Is Nothing Then
		Throw New ArgumentNullException("value")
	End If
	If startIndex < 0 Then
		Throw New ArgumentOutOfRangeException("startIndex", Environment.GetResourceString("ArgumentOutOfRange_NeedNonNegNum"))
	End If
	If startIndex > value.Length - 1 Then
		Throw New ArgumentOutOfRangeException("startIndex", Environment.GetResourceString("ArgumentOutOfRange_Index"))
	End If
	Contract.EndContractBlock()

	Return If((value(startIndex) = 0), False, True)
End Function
```
http://referencesource.microsoft.com/#mscorlib/system/bitconverter.cs,2af546f79a0b33fc

**Comment by Bill-McC** [comment]
Yep, so when you call BitConverter.GetBytes(int) and pass it to ToBoolean, it is effectively bitmasking the first byte, `value And &HFF <> 0`
The problem with that is you get the bizarre cases where 256 is False, but 258 is True.
The testing for zero is correct, it is the bitmasking that isn't. The definition for Boolean should be zero | non zero.

**Comment by Nukepayload2** [comment]
@Bill-McC you are right. The BitConverter is not suitable to implement the CVal operator. 
If the type of the first operand of CVal is Boolean, or the target type is Boolean, CVal shall be the same as CType.

**Comment by Bill-McC** [comment]
@Nukepayload2 I'm also for the same behaviour as CType . Be aware though that means CType(True, Int32) returns -1. For a Byte it is 255.  CType (Boolean, <Numeric>) returns 0 or it's bitwise compliment. For Double, Single etc, it is assumed a conversion to Int32 is done and then converted to the floating point type.
I know there are some bits of code out there that rely on the all bits set for tight loop bitsetting, but not sure how significant the cases are.  Considering CVal is all about performance, I think the decision to go with the VB all bits world, or the C# one bit set world, should be based on performance. If there is no significant difference, then I'd say keep it as CType.
 I don't think it would be wise to push the Boolean value straight into an int as that is easily corruptible, so the usual compare to zero is likely to have similar performance either way (I think).
